File: src/com/young/common/Singleton.java
package com.young.common;

/**
 * Author: taylorcyang
 * Date:   2014-10-21
 * Time:   15:58
 * Life with passion. Code with creativity!
 */
public abstract class Singleton<T> {
    private T mInstance;

    public abstract T create();

    public T get() {
        if(mInstance == null) {
            synchronized (this) {
                if (mInstance == null) {
                    mInstance = create();
                }
            }
        }
        return mInstance;
    }
}

File: app/src/main/java/com/young/common/Singleton.kt
package com.young.common

/**
 * Author: taylorcyang
 * Date:   2014-10-21
 * Time:   15:58
 * Life with passion. Code with creativity!
 */
abstract class Singleton<T> {

    @Volatile
    private var mInstance: T? = null

    abstract fun init(): T

    fun get(): T {
        val checkedInstance = mInstance
        if (checkedInstance != null) {
            return checkedInstance
        }
        return synchronized(this) {
            val checkInstanceAgain = mInstance
            if (checkInstanceAgain != null) {
                checkInstanceAgain
            } else {
                val created = init()
                mInstance = created
                created
            }
        }
    }
}
